tests/framework-tests/core/js/FluidIoCTests.js

```diff
@@ -5329,6 +5339,61 @@ https://github.com/fluid-project/infusion/raw/master/Infusion-LICENSE.txt
         that.destroy();
     });
 
+    /** Test to check for double inversion of notification order with onCreate createOnEvent **/
+
+    fluid.defaults("fluid.tests.FLUID6148notate", {
+        gradeNames: "fluid.component",
+        listeners: {
+            "onCreate.note": {
+                funcName: "fluid.tests.FLUID6148note",
+                args: ["{fluid.tests.FLUID6148create}", "{that}"],
+                // Note that if this priority is not supplied, the order will not be as expected since component
+                // creation via "onCreate" itself will not be fully enqueued
+                priority: "after:fluid-componentConstruction"
```

Comment:
The createOnEvent system isn't magic and has to be honoured by an actual, concrete listener added to "onCreate" - which has to take its natural place and priority with respect to the other ones. Simply adding any old "onCreate" listener to check for the creation point of the component runs the risk of racing against the framework's one.
This is a case where there has also been a change in behaviour between the old framework and the new one. The old framework did indeed use "any old listener", with respect to which the timing of another listener would be largely random - but on the other hand, guaranteed that the subcomponent would be FULLY constructed if you made your observation after the framework's one. Given that a major goal of this rewrite was to enable mutual reference between createOnEvent components created off the same event, the new framework now applies the creation in two stages. Firstly there is the same ordinary listener which starts the subcomponent creation process - but now, crucially, the framework also opens a transaction at this point, waiting for any other subcomponent creations kicked off by other ordinary listeners. Whilst this transaction is open, the subcomponents sit in a partially constructed state.
The framework then supplies a special, synthetic listener with priority "last:fluid-componentConstruction" which closes the transaction and kicks off the complete construction of all the createOnEvent subcomponents as a unit, which then allows them to refer to one another in their options.

This system is indeed rather over-complicated and arcane, but it is the only way out of a tight design corner bridging between the rather procedural semantics of events and the declarative semantics of component reference - without it, people had to make annoying redesigns of their component tree so that all the createOnEvent components had a single ancestor. This is another reason that we plan for createOnEvent components to be disused, now that we have finally implemented lensed components.

Reply:
Thanks for the explanation.